### ISSUE TITLE
strictFunctionTypes flag not checking class/interface methods

### ISSUE THREAD

**Issue by enylin** [microsoft/TypeScript#51094]
# Bug Report

### 🔎 Search Terms

contravariance, strictFunctionTypes, Function Parameter Bivariance, override, inheritance

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about substitutability, methods, and overloads

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.4#code/C4TwDgpgBAglC8UDeAoKUCGAuKBnYATgJYB2A5igL4oqiSwBCCsUAZMmlAEY77HlUaAeiFQAthGAALAPYATKBAAeGMWAA2EGqWAQCAMwwBjaAEkO6fQAowOGAEocANxlE5glEfUZcuKAGEYJiI1TQkSYD9zVEsbOwZ7ZCgRKCJgPFkAV3UFXFkAd0UCAhkCKCk9CAAabkz0khlpUjIoOSJcDQwQCHd0dCMZElwZTQA6dRkyKwByDC4EaZqwUa5xiHJpe05qak9B-FScc0QSCELAhistolHrJEwcaYBGACYAZmmoSkSUgkyIkLQPQlMoVAhaFApfT-IzAIiDKB0aBgEqQAigRQqUIQnR6QwmKCmF4WKD6HBxWCJeAAPigLjcHi8Pj8F2JIQ0EHCkUJxJipIpQUS9xSAzURE0ZWBpXKlU4-X2Iwg40mMzmCyWKzWGykW3QOxoQA)

### 💻 Code

```ts
type A = {
  a: string
}

type AB = A & {
  b: string
}

// method example

interface I {
  f(p: A): void
}

class CAB implements I {
  f(p: AB) { // it should show error here, but nothing displayed
    console.log('ab =', p.b.length)
  }
}

const i: I = new CAB()
i.f({ a: '123' }) // runtime error here

// function type property example

interface I2 {
  f: (p: A) => void
}

class CAB2 implements I2 {
  f(p: AB) { // compiler error here
    console.log('ab =', p.b.length)
  }
}
```

### 🙁 Actual behavior

The program crashed with the error message `Cannot read properties of undefined (reading 'length')`.
`strictFunctionTypes` flag works only when using class properties.

### 🙂 Expected behavior

`strictFunctionTypes` should also check interface methods.

**Comment by fatcerberus** [comment]
That's expected - methods are bivariant by design.  See https://github.com/microsoft/TypeScript/issues/49803#issuecomment-1176252140.

Per #18654 (the PR that implemented the flag):

> Methods are excluded specifically to ensure generic classes and interfaces (such as `Array<T>`) continue to mostly relate covariantly. The impact of strictly checking methods would be a much bigger breaking change as a large number of generic types would become invariant (even so, we may continue to explore this stricter mode).
